libmamba/src/solv-cpp/queue.cpp

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+
+#include <cassert>
+#include <limits>
+#include <exception>
+#include <sstream>
+
+#include <solv/queue.h>
+
+#include "solv-cpp/queue.hpp"
+
+namespace mamba::solv
+{
+    ObjQueue::ObjQueue(std::nullptr_t)
+        : m_queue{}
+    {
+    }
+
+    ObjQueue::ObjQueue()
+        : ObjQueue(nullptr)
+    {
+        queue_init(get());
+    }
+
+    ObjQueue::ObjQueue(std::initializer_list<value_type> elems)
+        : ObjQueue(elems.begin(), elems.end())
+    {
+    }
+
+    ObjQueue::ObjQueue(ObjQueue&& other)
+        : ObjQueue(nullptr)
+    {
+        swap(*this, other);
+    }
+
+    ObjQueue::ObjQueue(ObjQueue const& other)
+        : ObjQueue(other.begin(), other.end())
+    {
+    }
+
+    ObjQueue::~ObjQueue()
+    {
+        if (data() != nullptr)
+        {
+            queue_free(get());
+        }
+    }
+
+    auto ObjQueue::operator=(ObjQueue&& other) -> ObjQueue&
+    {
+        swap(*this, other);
+        // Leaving other empty to make sure ressources are no longer used
+        auto empty = ObjQueue(nullptr);
+        swap(other, empty);
+        return *this;
+    }
+
+    auto ObjQueue::operator=(ObjQueue const& other) -> ObjQueue&
+    {
+        auto other_copy = ObjQueue(other);
+        swap(*this, other_copy);
+        return *this;
+    }
+
+    void ObjQueue::push_back(value_type id)
+    {
+        queue_push(get(), id);
+    }
+
+    void ObjQueue::push_back(value_type id1, value_type id2)
+    {
+        queue_push2(get(), id1, id2);
+    }
+
+    auto ObjQueue::insert(const_iterator pos, value_type id) -> iterator
+    {
+        auto const offset = offset_of(pos);
+        insert(offset, id);
+        return begin() + offset;
+    }
+
+    auto ObjQueue::capacity() const -> size_type
+    {
+        return static_cast<size_type>(m_queue.count + m_queue.left);
+    }
+
+    auto ObjQueue::size() const -> size_type
+    {
+        assert(m_queue.count >= 0);
+        return static_cast<std::size_t>(m_queue.count);
+    }
+
+    auto ObjQueue::empty() const -> bool
+    {
+        return size() == 0;
+    }
+
+    auto ObjQueue::erase(const_iterator pos) -> iterator
+    {
+        auto const offset = offset_of(pos);
+        assert(offset <= std::numeric_limits<int>::max());
```

Comment:
Why `int` and not `std::ptrdiff_t`?
Same question in `reserve()`

Reply:
Because `int` is the type used by the `libsolv` functions `queue_delete`, `queue_prealloc`...